### PR TITLE
Support aria-roledescription in Edge.

### LINKED ISSUE #8448: UIA/Edge: no aria-roledscription and other properties are rendered in browse mode
Hi,

At the moment NVDA does not render aria-roledescription and friends via Microsoft Edge.

### Steps to reproduce:

1. With multiple web browsers open (including Edge), visit http://www.ets-research.org/ia11ylab/aria11spec/.
2. Navigate through examples. The key example to look at will be 29 and 30.
3. Use browse mode commands (including arrow keys) to read examples (in this case, buttons).

### Actual behavior:
NVDA does not announce "attachment button".

### Expected behavior:
NVDA announces "attachment button" along with button label if present.

#### NVDA Installed/portable/running from source:
Installed

#### NVDA version:
Next.15391

#### Windows version:
Windows 10 Version 1803

#### Name and version of other software in use when reproducing the issue:
Microsoft Edge, Firefox 61.0

#### Other information about your system:
Windows 10 App Essentials add-on is installed, but the result is the same no matter if add-ons are enabled or disabled.

#### Does the issue still occur after restarting your PC?
Yes

#### Have you tried any other versions of NVDA?
Yes (latest master, 2018.2.1)

### Technical details

When using the following procedure, roledescriptoin and friends are returned:

1. With Edge opened to the website described above, focus on the attachment button.
2. Open Python Console (Control+NvDA+Z) and do:

import UIAHandler
nav._getUIACacheablePropertyValue(30102)

Note: 30102 is UIA_AriaPropertiesPropertyId constant.

If properties exist, they will appear as a string value of the form:

property=str;property2=str; ...

Which should then allow str.split(";") to return a list of property=str collection, which can be coerced into a properties dictionary. This dictionary should then be used to work with these properties in various places, chiefly in UIA browse mode module.

Thanks.

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #8448 

### Summary of the issue:
Web authors can provide an aria-roledescription property to customize the way the type (role) of an element is announced/brailled in Assistive technologies. NVDA supports this already for Chrome/Firefox/IE but not Edge.

### Description of how this pull request fixes the issue:
Implements support for aria-roledescription in Edge for both focus and browse mode.
This PR also now:
* Ensures that braille honors roledescription in browseMode (it already did in focus mode).
* Ensures that the name of a field is not duplicated if the field is also a landmark.
* Region is no longer reported for a landmark region that is also another valid field (E.g. a region with a roledescription).
* roledescription on divs in Edge now present as block fields (I.e. do not keep reporting the roledescription for each line within the field).

### Testing performed:
Tested example from #8448 in Edge.

### Known issues with pull request:
None.

### Change log entry:
Existing what's new for aria-roledescription should be updated to mention Edge, if this is merged before 2018.3.
Section: New features, Changes, Bug fixes

### CODE REVIEW

**Comment by josephsl** [comment]
Hi,

Testing shows it is working as expected. I may need to test it on multiple Windows 10 releases to see which one added this property for Edge nodes.

Thanks.

**Review by feerrenrut — DISMISSED** [review]

**Comment by leonardder** [comment]
I tested some of the examples from [the ARIA examples page](www.ets-research.org/ia11ylab/aria11spec/) and found the following.

`<div  role="region" aria-roledescription="slide" id="slide42" aria-labelledby="slide42heading">``
+ Expected: Quarterly Report slide
+ Actual: Quarterly Report region slide

Note that this also is the case in firefox. In edge, the section variant of slide reports correctly, whereas in edge it still reports region redundantly.

Another thing you might be able to fix as part of this, is that roletext seems to be entirely ignored for braille.

**Comment by michaelDCurran** [comment]
I believe I have now addressed the most recent issues.
New changes:
* Ensures that braille honors roledescription in browseMode (it already did in focus mode).
* Ensures that the name of a field is not duplicated if the field is also a landmark.
* Region is no longer reported for a landmark region that is also another valid field (E.g. a region with a roledescription).
* roledescription on divs in Edge now present as block fields (I.e. do not keep reporting the roledescription for each line within the field).

File: source/NVDAObjects/UIA/edge.py [hunk 1]
```diff
@@ -133,6 +133,13 @@ def move(self,unit,direction,endPoint=None,skipReplacedContent=True):
 	def _getControlFieldForObject(self,obj,isEmbedded=False,startOfNode=False,endOfNode=False):
 		field=super(EdgeTextInfo,self)._getControlFieldForObject(obj,isEmbedded=isEmbedded,startOfNode=startOfNode,endOfNode=endOfNode)
 		field['embedded']=isEmbedded
+		role=field.get('role')
+		# Fields should be treeted as block for certain roles.
```
Comment (leonardder) [inline source/NVDAObjects/UIA/edge.py]: treeted > treated

File: source/browseMode.py [hunk 2]
```diff
@@ -1046,30 +1046,30 @@ def getControlFieldSpeech(self, attrs, ancestorAttrs, fieldType, formatConfig=No
...
+				if landmark == "region":
+					# The word landmark is superfluous for regions.
+					textList.append(aria.landmarkRoles[landmark])
+			if landmark != "region":
 				textList.append(_("%s landmark") % aria.landmarkRoles[landmark])
 		textList.append(super(BrowseModeDocumentTextInfo, self).getControlFieldSpeech(attrs, ancestorAttrs, fieldType, formatConfig, extraDetail, reason))
 		return " ".join(textList)
 
-	def getControlFieldBraille(self, field, ancestors, reportStart, formatConfig):
+	def getControlFieldBraille(self, field, ancestors, reportStart, formatConfig,reason=None):
```
Comment (leonardder) [inline source/browseMode.py]: Is this extra reason kwarg ever used and different than None? I'd say you can also just pass reason=None to field.getPresentationCategory

**Review by LeonarddeR — COMMENTED** [review]
I will do some testing with braille later today.

**Review by LeonarddeR — APPROVED** [review]
This looks ok to me, also from my braille display :)

There is one major thing I found, namely that by default, roles don't seem to show up in Edge browse mode at all, unless you're using quick nav. However, that is completely unrelated to this issue. Just making note of it for future reference.
